### ISSUE TITLE
Do you release Tubefeeder under the GPLv3-only or the GPLv3-or-later?

### ISSUE THREAD

**Issue by EmanuelLoos** [Schmiddiii/Tubefeeder#28]
In case you haven't decided yet, I would recommend the GPLv3-or-later under which, for example, NewPipe is also licensed. Otherwise it can get very difficult or almost impossible to use your code under a newer version of the GPL when there are many contributions since every single contributor would have to be asked again and agree to that.

**Comment by Schmiddiii** [comment]
I have not thought about that at all. If you recommend it, it will be licensed under GPLv3-or-later. Do I have to change out the license for that or how do I show the "or-later" on github?

**Comment by EmanuelLoos** [comment]
> If you recommend it, it will be licensed under GPLv3-or-later. Do I have to change out the license for that or how do I show the "or-later" on github?

In order to define that you allow your software to be used under the terms of the GNU General Public License version 3 or any later version of the GPL you wold need a copyright notice often named "NOTICE" or in NewPipes case "[copyright](https://github.com/TeamNewPipe/NewPipe/blob/dev/copyright)" stating that. [The GNU project also recommends adding a short statement about the license of each file following a copyright notice](http://www.gnu.org/licenses/gpl-howto.html). [Here you can see how it is done in NewPipes source code](https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/player/MainPlayer.java).

If you are interested about the or-later clause you can also look [here](http://www.gnu.org/licenses/gpl-faq.html#VersionThreeOrLater).

**Comment by Schmiddiii** [comment]
I have made a quick pull request regarding this issue. Might you please check if everything seems ok?

**Comment by EmanuelLoos** [comment]
Lokking good, thoug you forgot to change Foobar to Tubefeeder in one line:

>You should have received a copy of the GNU General Public License
along with Foobar.  If not, see <https://www.gnu.org/licenses/>.

**Comment by EmanuelLoos** [comment]
Oh, and Tubefeeder should also reproduce the copyright notice and the license (about section in the GUI).

**Comment by Schmiddiii** [comment]
I have added a about section in the mentioned pull request. Please look at it if you can.

The current weird thing I currently do not know how to solve is that it requires two clicks of the back-button on the about page to exit it. Might be my fault or something wrong with gtk.

**Comment by EmanuelLoos** [comment]
I am a little lost building Tubefeeder on my Devuan (Debian without systemd) system so I didn't manage to review the about section yet, but when trying to build your project it also checks out [this](https://github.com/Schmiddiii/file-minidb) repository which doesn't have a copyright and license notice stating releasing it under GPLv3-or-later. If this is compiled with Tubefeeder I would recommend you to add a notice there too for it to be GPLv3-or-later for the compiled binary to be GPLv3-or-later. Otherwise it would make the compiled binary GPLv3-only (GPLv3-or-later + GPLv3-only = GPLv3-only).

**Comment by Schmiddiii** [comment]
This repository is mine, so I can include a notice. I will probably drop this dependency on the next version so I do not think this will be a big problem.

Regarding the compilation issues, is there anything I can help with?

**Comment by EmanuelLoos** [comment]
> Regarding the compilation issues, is there anything I can help with?

Thanks, in the end `cargo build` worked, no `cross` command on my system (maybe I am missing some dependencies).

With Docker I ran out of storage space. What is it needed for in the build process?

**Comment by Schmiddiii** [comment]
Cross is not needed when building for the PC you are on, just for building for th e PinePhone.

I have not checked yet how much space compilation is uaing, but rust compilation takes a lot of space in general (just look at the `target` folder).

Have you already looked at the about dialog. Does it seem correct?

**Comment by EmanuelLoos** [comment]
> Have you already looked at the about dialog. Does it seem correct?

I think so, it seems just like the about dialogs in other GTK programs. However, I would recommend a button to view the license text locally as well because URLs can change over time. You probably also want to mention if Tubefeeders name and icon may or may not be used in forks (if not then I would recommend an exception for forks that are not really a new project, a new program that is distributed everywhere but are just used to develop a new feature to later make a pull request) and for the AppImage and Flatpack I would recommend to also mention the included dependencies and their licenses.

**Comment by Schmiddiii** [comment]
> I think so, it seems just like the about dialogs in other GTK programs.

It is the same. There is a [about dialog included in gtk](https://docs.gtk.org/gtk3/class.AboutDialog.html).

> However, I would recommend a button to view the license text locally as well because URLs can change over time.

I do not want to change the default about menu too much. As far as I see I cannot show [license-type](https://docs.gtk.org/gtk3/property.AboutDialog.license-type.html) and [license](https://docs.gtk.org/gtk3/property.AboutDialog.license.html) at the same time, so I do not think this is easily possible.

> You probably also want to mention if Tubefeeders name and icon may or may not be used in forks.

Forks can use the name and icon if they want. Where should I put this information. This would probably not fit in the [about dialog](https://docs.gtk.org/gtk3/class.AboutDialog.html) as there is no such thing as additional data.

> AppImage and Flatpack I would recommend to also mention the included dependencies and their licenses.

Would not fit into the [about dialog](https://docs.gtk.org/gtk3/class.AboutDialog.html) for the same reason. Furthermore searching for all the dependencies and recursive dependencies seems hard to do. As long as it is not required I think the efford would not be worth doing it.

**Comment by EmanuelLoos** [comment]
> I do not want to change the default about menu too much. As far as I see I cannot show license-type and license at the same time, so I do not think this is easily possible.

You are probably right. Doing it like the other programs should be enough.

> Forks can use the name and icon if they want. Where should I put this information. This would probably not fit in the about dialog as there is no such thing as additional data.

After looking a bit around I think unless Tubefeeder is a registered trademark you don't have to do anything. (It is just very common even for free software projects to not allow forks to use the same name and icons.) You more or less licensed the icon under the GPLv3 by putting it in this GPLv3-licensed repository though just like with the code I think it would not hurt to specify the license in its metadata too. The GPL is mainly intended for source code though so maybe you would like to allow its usage under the terms of the GNU FDL (documentations) and CC-BY-SA-4.0 (documentations, wikis, icon packs) as well as under the terms of the GPLv3. By the way under which license do you release the documentation on wiki?

> Would not fit into the about dialog for the same reason. Furthermore searching for all the dependencies and recursive dependencies seems hard to do. As long as it is not required I think the efford would not be worth doing it.

I thought about a "Dependencies included in the AppImage" section in the credits **only listing dependencies included in it**. Most free software licenses including the GPL, the [MIT license](https://en.wikipedia.org/wiki/MIT_License) (though the MIT license only requires inclusion, not reproduction of the license in the UI) or some [BSD licenses](https://en.wikipedia.org/wiki/BSD_licenses) require attribution. It doesn't go against the [four essential freedoms](https://www.gnu.org/philosophy/free-sw.html#fs-definition) so from the free software point of view it is no problem. Because of that free software often comes from schools or universities as their goal is not to make money with it directly (which is normally done by charging money for its use going against the four essential freedoms) but to let the world know what a great project they made so people choose to study there, they want it to be used as much as possible with attribution. A Debian package as well as packages on other GNU/Linux distributions I know doesn't contain its dependencies, they are resolved by apt and the user is asked if he/she wants to install listed dependencies. As the package doesn't contain these (often there are also multiple possibilities) its UI doesn't have to attribute them, the package manager shows them instead (and shows additional information about them on the users request). An Android App on the other hand contains everything it needs which is why you may often find an "Open Source Licenses" section in them attributing everything included. An AppImage also includes dependencies as far as I know so I do not see any difference. However, I am not a lawyer even though I might know quite much about the idea of free software and copyleft.

**Comment by Schmiddiii** [comment]
> The GPL is mainly intended for source code though so maybe you would like to allow its usage under the terms of the GNU FDL (documentations) and CC-BY-SA-4.0 (documentations, wikis, icon packs) as well as under the terms of the GPLv3.

Where should I put a disclaimer to license these things under CC, some things are .png or similar and can not have a notice included.

> By the way under which license do you release the documentation on wiki?

Have not thought about that. Maybe CC like the other metadata mentioned above?.
How would I specify this, maybe a extra LICENSE page?

> I thought about a "Dependencies included in the AppImage" section in the credits only listing dependencies included in it. Most free software licenses including the GPL, the MIT license (though the MIT license only requires inclusion, not necessarily reproduction of the copyright notice and license in the UI) or some BSD licenses require attribution. It doesn't go against the four essential freedoms so from the free software point of view it is no problem. Because of that free software often comes from schools or universities as their goal is not to make money with it directly (which is normally done by charging money for its use going against the four essential freedoms) but to let the world know what a great project they made so people choose to study there, they want it to be used as much as possible with attribution. A Debian package as well as packages on other GNU/Linux distributions I know doesn't contain its dependencies, they are resolved by apt and the user is asked if he/she wants to install listed dependencies. As the package doesn't contain these (often there are also multiple possibilities) its UI doesn't have to attribute them, the package manager shows them instead (and shows additional information about them on the users request). An Android App on the other hand contains everything it needs which is why you may often find an "Open Source Licenses" section in them attributing everything included. An AppImage also includes dependencies as far as I know so I do not see any difference. However, I am not a lawyer even though I might know quite much about the idea of free software and copyleft.

I do not really understand if it is required or not. Searching around flathub on a few flatpaks (not too many) I am using, I did not see any software including license information of its dependencies.

**Comment by EmanuelLoos** [comment]
> Where should I put a disclaimer to license these things under CC, some things are .png or similar and can not have a notice included.

I thought about including the notice of the PNG in the EXIF metadata of the PNG (just a short notice, not the whole license) and if you would like to allow its usage under the terms of the CC-BY-SA-4.0 and the GNU FDL as well as under the terms of the GPLv3-or-later also stating something like "In addition, as a special exception, you may also (at your opinion) use the icon of Tubefeeder located at packaging/Tubefeeder.png under the terms of the GNU FDL, either version 1.3 (https://www.gnu.org/licenses/fdl-1.3.html) or (at your opinion) any later version of the license, or the CC-BY-SA-4.0-license (https://creativecommons.org/licenses/by-sa/4.0/legalcode)." in the license notice (EXIF metadata + NOTICE file in the root directory).

Note that it is always important to specify which [CC license](https://en.wikipedia.org/wiki/Creative_Commons_license#Types_of_license) you mean as only the "free culture licenses" of them grant the for essential freedoms and the others are non-free.

> Have not thought about that. Maybe CC like the other metadata mentioned above?.
How would I specify this, maybe a extra LICENSE page?

I think an extra LICENSE page will do. Maybe also a custom footer stating something like "Unless otherwise noted content on this wiki may be used under the terms of the GNU FDL, either version 1.3 (https://www.gnu.org/licenses/fdl-1.3.html) or (at your opinion) any later version of the license, or (at your opinion) the CC-BY-SA-4.0-license (https://creativecommons.org/licenses/by-sa/4.0/legalcode)."

> I do not really understand if it is required or not. Searching around flathub on a few flatpaks (not too many) I am using, I did not see any software including license information of its dependencies.

Do flatpaks include their dependencies or does the flatpak package manager handle them? As far as I know AppImages  include their dependencies which means when distributing an AppImage you also distribute its dependencies with it. The Tubefeeder GUI is kind of the UI of the AppImage. It depends on the licenses of the included dependencies and how you interpret them if you need to attribute them or not. In any way most licenses, even permissive licenses, require you to include the original copyright notice (who is the copyright holder?), license notice (under which license is which part of the project distributed?, are there any special exceptions?) and license (the actual license text) when distributing the project. Some of them require interactive user interfaces to also display this information. You can see how this is done in Android Applications like NewPipe (About --> Licenses). Android Apps are like containers distributed with all their dependencies (except for the Android OS, of course). If the same is the case for AppImages I think the same applies here. However, a package that doesn't include its dependencies because they are meant to be resolved by the package manager doesn't have to do anything about the licenses of its dependencies because the package manager does that (which is the case for most GTK programs, I think).

**Comment by Schmiddiii** [comment]
Flatpak ships with the dependencies, similar to AppImage. 

AppImage only ships with some libraries (see [here](https://github.com/Schmiddiii/Tubefeeder/blob/3911d65fe97182d608d2923b995d1fb88af1fe4b/packaging/AppImage/AppImageBuilder.yml#L31)), gtk, youtube-dl, mpv and other dependencies are taken from the system.

[Flatpak](https://github.com/flathub/de.schmidhuberj.tubefeeder) also includes mpv and youtube-dl.

Note that these probably also have sub-dependencies that might need to be checked.

I believe that both already include the licenses and notices in the package (AppImage simulates a installation using apt, Flatpak builds these and they will put the licenses into the right place). They are not easily accessable, but can be extracted.

**Comment by EmanuelLoos** [comment]
> They are not easily accessable, but can be extracted.

That seems a bit like a grey zone of interpretation. I think most people will take an Android App, a Flatpak or an AppImage as one executable with one UI, rather then an archive containing a bunch of files they could extract, read and run too. At least for Android Apps the UI of the App is normally expected to mention all licenses of included software. The UI is kind of seen as the one UI of the App and the App is seen as one program.

**Comment by Schmiddiii** [comment]
I have forwarded the question about including the licenses in the UI of the application to the [flatpak matrix room](https://matrix.to/#/#flatpak:matrix.org). Maybe they have a answer.

**Comment by Schmiddiii** [comment]
I have now added a exiv comment to tubefeeder.png (readable for example with [exiv2](https://linux.die.net/man/1/exiv2)) and updated the NOTICE file to include the exception.

The people at the flatpak matrix room do not think that it is neccessary to add a gui-element for third-party dependencies.

I believe I only have to add a LICENSE to the the wiki and merge the pull request, then this issue can be closed, is this right?

**Comment by EmanuelLoos** [comment]
Looks good, what also came to my mind: 

If you would like to accept contributions it would be a good idea to have something like a "I confirm that this is either my code or was released under the terms of a GPLv3-or-later compatible license. Also I agree to release it here under the terms of the GPLv3-or-later." checkbox in a pull request template. Same would probably be a good idea for the wiki with CC-BY-SA-4.0. That way licensing conflicts could be avoided.

**Comment by Schmiddiii** [comment]
I have now added a PR-template. Sadly this can only be tested if the branch got merged into master.

Regarding the wiki I have added a section [Contributing](https://github.com/Schmiddiii/Tubefeeder/wiki#contributing) to the wiki as there are no PR-templates (as far as I know) for wikis.

**Comment by Schmiddiii** [comment]
I am going to merge this. If anything else is left to do, please reopen.

**Comment by Schmiddiii** [comment]
> a) the GNU FDL (any version) with the or-later clause (the user can choose this version or any later version of it) as well as the CC-BY-SA-4.0 license,

Why now `any version` of the license? I thought only 1.3 or later.

> Though I am not sure a contributing section is enough

It should be enough as it is clearly visible. A checkbox on a pull request would of course be better, but as wikis currently do not support such things, it will have to be good enough.

> If it can't handle licensing to well you might want to take a look at MediaWiki

Yes, Github it cannot handle licensing of wikis in a nice way. But I will still keep the wiki on Github as it provides a central place for the entire wiki. What I might do is move the wiki to a new repository and make it visible using github pages, this will make licensing more viable, but as the wiki is not too large I do not think this will be neccessary for now. Unless Github goes crazy some day I will not move to MediaWiki or similar technology as I do not want my documentation and code in two different places.

**Comment by EmanuelLoos** [comment]
> Why now any version of the license? I thought only 1.3 or later.

Any version of the GNU FDL with the "or ony later version" clause can be turned into the GNU FDL version 1.3 or-later.

Including something from another source and relicensing it to GFDLv13-or-later and CC-BY-SA-4.0 (for the licensee to choose one of them) requires the original work to be licensed in a way that allows both licenses (not only one). Then, anyone wishing to use content on this wiki can choose the one that works for them. You are offering multiple copyleft licenses not too much compatible with each other which is the reason for offering both. In order to do that, however, the original work needs to be licensed in a way that allows offering all of these. The original license could either be a permissive license (like CC0 [though for software one needs to be careful about patents with this license], MIT license, etc) or a copyleft license that is compatible with all offered licenses. Copyleft (the word has no legal relevance, "Copyleft <name>" is not legally valid) is copyright turned against itself partly to disallow misuse of it like taking your code to make and sell non-free software out of it, like DRM, etc. In short, copyleft is there to stop peoples rights being restricted based on copyright that has become too extreme nowadays (especially around software). Therefore copyleft is difficult: Licenses are incompatible and code needs to be rewritten etc. Because of this it might be advisable to offer multiple licenses (while it still can be done). You can go back to less copyleft licenses from which the licensee can choose from always (for further changes, a free license can't be revoked) but not the other way around.

> It should be enough as it is clearly visible. A checkbox on a pull request would of course be better, but as wikis currently do not support such things, it will have to be good enough.

A license is a legal document and clicking "I agree", ticking a checkbox, etc is like signing that document. Without the original copyright owner expressing that he/she licenses something under the terms of some license to someone it is very questionable if that is a valid license. This is almost the very reason youtube-dl can legally exist as the YouTube terms of service are liked on YouTube but only agreed to when signing up (though it is also questionable how far some terms of service can restrict the legal right to create private copies of copyrighted works [only applies as long as the template (the work being copied) is not distributed in a obviously illegal manner]).

**Comment by Schmiddiii** [comment]
> Including something from another source

I believe everything on this wiki was written by myself or @calebccff, so I do not think we have other sources.

> A license is a legal document and clicking "I agree", ticking a checkbox, etc is like signing that document.

Yes, I agree, but the only alternative I currently see is making the wiki only editable by me and rewriting @calebccff part of the wiki. That would guarantee that everything is fine with the licenses. I really do not want to do this, so the `Contributing`-Section seems to be the best alternative. I could also extract the wiki into its own repository and using github pages as the wiki, but this will need time I currently am not willing to invest.

**Comment by EmanuelLoos** [comment]
> I believe everything on this wiki was written by myself or @calebccff, so I do not think we have other sources.

My point. Right now there aren't other sources so it can easily be dual licensed this way. "By contributing to this wiki ..." is stating that a own work is allowed to be included if the contributor agrees to release it under these licenses (GFDLv1.3 and CC-BY-SA-4.0) and is also stating on which conditions external content could be included and relicensed.

> Yes, I agree, but the only alternative I currently see is making the wiki only editable by me and rewriting @calebccff part of the wiki.

With only one contributor this is doable without rewriting: You would need a written consent from @calebccff to distribute calebccffs contributions under the terms of the GFDLv1.3 and to distribute them under the terms of the CC-BY-SA-4.0. However, this gets more and more complex with more contributors. Especially if one of them is not reachable anymore.

> the Contributing-Section seems to be the best alternative

I am not sure this alternative works legally because in order to license something you need the copyright owner to express this: "I agree to ...". "By doing this you agree to that" might work if is is visible clicking where the button "do that" is. But if someone isn't on the homepage of the wiki but somewhere else the "By doing this you agree to that" is not visible when clicking on "Edit" or "New Page".

**Comment by Schmiddiii** [comment]
I have now made the wiki editable only by contributors and contacted caleb about the relicense. This is definitly not a ideal solution, so I created a new card in the [contributions project](https://github.com/Schmiddiii/Tubefeeder/projects/3) extracting the wiki to its own repository. In the new repository we can then make a PR-template to for licensing. As I currently want to do other things, this is open for everyone to do.

**Comment by Schmiddiii** [comment]
I have added more information on what to do in the card, including creating the license notice. If you think this can be closed now, please close otherwise just leave it open

**Comment by EmanuelLoos** [comment]
Looks good.

**Comment by Schmiddiii** [comment]
I have now started to move the wiki to its own [repository](https://github.com/Tubefeeder/wiki/tree/gh-pages), the wiki is available [here](https://tubefeeder.github.io/wiki/). This repository has a pull request template with 

___
I confirm that this is either my code or was released under the terms of a GNU FDL 1.3 or later compatible license and the CC-BY-SA-4.0 license. Also I agree to release it here under the terms of the GNU FDL 1.3 or later and the CC-BY-SA-4.0 license.
___

as a checkbox, i hope this should be enought. I will probably finalize the move in a few days, but the url might change as I do not like the `github.io` part. Maybe I will buy a `tubefeeder`-domain.

**Comment by EmanuelLoos** [comment]
Looks good!
Though there are a few more things:
1. There is no "NOTICE" stating under which licenses the wiki is available.
How about the custom footer from earlier:
"Unless otherwise noted content on this wiki may be used under the terms of the GNU FDL, either version 1.3 (https://www.gnu.org/licenses/fdl-1.3.html) or (at your opinion) any later version of the license, or (at your opinion) the CC-BY-SA-4.0-license (https://creativecommons.org/licenses/by-sa/4.0/legalcode)."
2. The copyright and license notice is not visible on the wiki. Could the "NOTICE" file somehow be shown at the end of each page of the wiki?
3. The "LICENSE" file is not actually the license of the wiki. There are two licenses anyone who wishes to use content on the wiki can choose from. Therefore I would rename it to something like "GFDLv1.3". (Actually, with this combination you don't have to include the GNU FDL because only the requirements of one license have to be fulfilled but you can include it, of course, if you like [Though I would include both licenses then, to avoid confusion])

**Comment by Schmiddiii** [comment]
1. I totally forgot that. It should now be fixed.
2. Every page should now have the contents of the NOTICE (including the copyright) on the bottom.
3. I renamed to LICENSE to LICENSE_GFDLv1.3. I believe the inclusion of CC-BY-SA-4.0 is not mandatory for the license and as I cannot currently find a official text or markdown file of this license I will leave it out.

**Comment by EmanuelLoos** [comment]
Haven't there already been contributions? In that case it isn't entirely your copyright so the copyright notice should state who's copyright it is. I have also often seen something general like adding "and contributors" after your name (if it is mostly your work) or simply writing "Copyright Tubefeeder contributors" though then I think it at least needs to be visible somewhere who the contributors (the copyright holders) are (for example a file with the names of the contributors).

3. I believe [this](https://creativecommons.org/licenses/by-sa/4.0/legalcode) contains the official license text. You are right that inclusion of the CC-BY-SA-4.0 license text is not mandatory, howerver, I think if only one is included it might be a bit confusing. Especially the button "View license" on the repo of the wiki which directly takes me to the GFDL mentioning nothing about the CC-BY-SA-4.0 license.

**Comment by Schmiddiii** [comment]
I have now included the CC-BY-SA-4.0 text from [here](https://creativecommons.org/licenses/by-sa/4.0/legalcode.txt) and modified the notice to be extendable to include the copyright of the authors. I have contacted the only other contributor so far if he wants to include his copyright. I have also updated the pull request template to include

~~~
I confirm that I have put copyright information in the new or edited files if wanted.
~~~

As far as I see in the [faq](https://www.gnu.org/licenses/gpl-faq.en.html#RequiredToClaimCopyright) of gpl, including a copyright is not neccessary if not wanted. I have not checked for CC-BY-SA-4.0 but I believe it will be similar.

**Comment by Schmiddiii** [comment]
I have now included the CC-BY-SA-4.0 text from [here](https://creativecommons.org/licenses/by-sa/4.0/legalcode.txt) and modified the notice to be extendable to include the copyright of the authors. I have contacted the only other contributor so far if he wants to include his copyright. I have also updated the pull request template to include

~~~
I confirm that I have put copyright information in the new or edited files if wanted.
~~~

As far as I see in the [faq](https://www.gnu.org/licenses/gpl-faq.en.html#RequiredToClaimCopyright) of gpl, including a copyright is not neccessary if not wanted. I have not checked for CC-BY-SA-4.0 but I believe it will be similar.
